test(types): restore TypeScript 5.9 to CI matrix

## Context

TypeScript 5.9 was provisionally added to the `test-types` CI matrix when expanding from `['5.0', '5.1', '5.2']` to broader coverage, but was dropped before merge because `pnpm test:ts` (`tsc --noEmit`) fails on it.

The repo currently develops against `5.7.2` and the matrix has been updated to `['5.0', '5.4', '5.7']`. This issue tracks restoring 5.9 once the failures below are addressed.

## Reproduction

```sh
pnpm install --frozen-lockfile
pnpm build-ci
pnpm add -w typescript@5.9
pnpm test:ts
```

`tsc --version` → `5.9.3`.

## Failure output

```
packages/core/src/helpers.ts(19,43): error TS2322: Type 'T' is not assignable to type 'T extends AnyFn<any, infer U> ? U : T'.
packages/core/src/helpers.ts(55,9): error TS2322: Type 'undefined' is not assignable to type 'T[P]'.
  'undefined' is assignable to the constraint of type 'T[P]', but 'T[P]' could be instantiated with a different subtype of constraint 'any'.
packages/core/src/helpers.ts(188,7): error TS2322: Type 'any[]' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to 'any[]'.
packages/shared/src/helpers.ts(71,15): error TS2322: Type 'never[]' is not assignable to type 'Arrify<Exclude<T, void>>'.
packages/shared/src/helpers.ts(71,45): error TS2322: Type 'T[]' is not assignable to type 'Arrify<Exclude<T, void>>'.
```

All five come from tightened conditional-type narrowing in 5.9 — the spring engine leans on `any` deliberately, so some of these are likely intentional escape hatches that 5.9 now refuses to infer through.

## Definition of done

- Failures resolved in source (preferred) or scoped behind narrowly-targeted `@ts-expect-error` annotations with a TS-5.9-specific comment.
- `pnpm test:ts` clean on 5.9.
- `5.9` added back to the `matrix.ts` array in `.github/workflows/tests.yml`.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -141,7 +141,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        ts: ['5.0', '5.4', '5.7']
+        ts: ['5.0', '5.4', '5.7', '5.9']
 
     steps:
       - name: Checkout repo

--- a/packages/core/src/helpers.ts
+++ b/packages/core/src/helpers.ts
@@ -12,10 +12,16 @@ import { ReservedProps, ForwardProps, InferTo } from './types'
 import type { Controller } from './Controller'
 import type { SpringRef } from './SpringRef'
 
+export function callProp<T extends AnyFn>(
+  value: T,
+  ...args: Parameters<T>
+): ReturnType<T>
+export function callProp<T>(value: Exclude<T, AnyFn>, ...args: unknown[]): T
 export function callProp<T>(
   value: T,
   ...args: T extends AnyFn ? Parameters<T> : unknown[]
-): T extends AnyFn<any, infer U> ? U : T {
+): T extends AnyFn<any, infer U> ? U : T
+export function callProp(value: unknown, ...args: unknown[]) {
   return is.fun(value) ? value(...args) : value
 }
 
@@ -47,7 +53,7 @@ export const hasDefaultProp = <T extends Lookup>(props: T, key: keyof T) =>
 export const getDefaultProp = <T extends Lookup, P extends keyof T>(
   props: T,
   key: P
-): T[P] =>
+): T[P] | undefined =>
   props.default === true
     ? props[key]
     : props.default
@@ -183,15 +189,17 @@ export function inferTo<T extends object>(props: T): InferTo<T> {
 
 // Compute the goal value, converting "red" to "rgba(255, 0, 0, 1)" in the process
 export function computeGoal<T>(value: T | FluidValue<T>): T {
-  value = getFluidValue(value)
-  return is.arr(value)
-    ? value.map(computeGoal)
-    : isAnimatedString(value)
-      ? (G.createStringInterpolator({
-          range: [0, 1],
-          output: [value, value] as any,
-        })(1) as any)
-      : value
+  const resolved = getFluidValue(value)
+  if (is.arr(resolved)) {
+    return resolved.map(computeGoal) as T
+  }
+  if (isAnimatedString(resolved)) {
+    return G.createStringInterpolator({
+      range: [0, 1],
+      output: [resolved, resolved],
+    })(1) as T
+  }
+  return resolved as T
 }
 
 export function hasProps(props: object) {

--- a/packages/shared/src/helpers.ts
+++ b/packages/shared/src/helpers.ts
@@ -68,7 +68,7 @@ export function eachProp<T extends object, This>(
 }
 
 export const toArray = <T>(a: T): Arrify<Exclude<T, void>> =>
-  is.und(a) ? [] : is.arr(a) ? (a as any) : [a]
+  (is.und(a) ? [] : is.arr(a) ? a : [a]) as Arrify<Exclude<T, void>>
 
 /** Copy the `queue`, then iterate it after the `queue` is cleared */
 export function flush<P, T>(

